Havtorn/Source/Engine/Graphics/RenderManager.cpp

```diff
@@ -813,6 +787,239 @@ namespace Havtorn
+			if (materialProperty.TextureChannelIndex > -1)
+			{
+				U16 textureIndexCopy = *textureIndex;
+				if (!textureIndices.contains(textureIndexCopy))
+				{
+					resourceViewPointers.emplace_back(textureBank->GetTexture(textureIndexCopy));
+					textureIndices.emplace(textureIndexCopy, static_cast<U16>(resourceViewPointers.size() - 1));
+				}
+
+				*textureIndex = textureIndices[textureIndexCopy];
+			}
+		};
+
+		findTextureByIndex(materialCopy.AlbedoR);
+		findTextureByIndex(materialCopy.AlbedoG);
+		findTextureByIndex(materialCopy.AlbedoB);
+		findTextureByIndex(materialCopy.AlbedoA);
+		findTextureByIndex(materialCopy.NormalX);
+		findTextureByIndex(materialCopy.NormalY);
+		findTextureByIndex(materialCopy.NormalZ);
+		findTextureByIndex(materialCopy.AmbientOcclusion);
+		findTextureByIndex(materialCopy.Metalness);
+		findTextureByIndex(materialCopy.Roughness);
+		findTextureByIndex(materialCopy.Emissive);
+
+		auto convertToFloats = [&](SEngineGraphicsMaterialProperty& materialProperty, int propertyIndex)
+		{
+			MaterialBufferData.Properties[propertyIndex].ConstantValue = materialProperty.ConstantValue;
+			MaterialBufferData.Properties[propertyIndex].TextureIndex = static_cast<F32>(materialProperty.TextureIndex);
+			MaterialBufferData.Properties[propertyIndex].TextureChannelIndex = static_cast<F32>(materialProperty.TextureChannelIndex);
+		};
+
+		convertToFloats(materialCopy.AlbedoR, 0);
+		convertToFloats(materialCopy.AlbedoG, 1);
+		convertToFloats(materialCopy.AlbedoB, 2);
+		convertToFloats(materialCopy.AlbedoA, 3);
+		convertToFloats(materialCopy.NormalX, 4);
+		convertToFloats(materialCopy.NormalY, 5);
+		convertToFloats(materialCopy.NormalZ, 6);
+		convertToFloats(materialCopy.AmbientOcclusion, 7);
+		convertToFloats(materialCopy.Metalness, 8);
+		convertToFloats(materialCopy.Roughness, 9);
+		convertToFloats(materialCopy.Emissive, 10);
+
+		MaterialBufferData.RecreateZ = materialCopy.RecreateNormalZ;
+		BindBuffer(MaterialBuffer, MaterialBufferData, "Material Buffer");
+
+		Context->PSSetShaderResources(5, static_cast<U32>(resourceViewPointers.size()), resourceViewPointers.data());
+		Context->PSSetConstantBuffers(8, 1, &MaterialBuffer);
+		
+		Context->DrawIndexed(static_cast<U32>(GeometryPrimitives::Icosphere.Indices.size()), 0, 0);
+		CRenderManager::NumberOfDrawCallsThisFrame++;
+
+		// ======== Lighting =========
+		Context->OMSetRenderTargets(1, &renderTarget, nullptr);
+
+		SEnvironmentLightComponent* environmentLightComp = new SEnvironmentLightComponent(tempEntity, EComponentType::EnvironmentLightComponent);
+		SDirectionalLightComponent* directionalLightComp = new SDirectionalLightComponent(tempEntity, EComponentType::DirectionalLightComponent);
+
+		LoadEnvironmentLightComponent("Skybox", environmentLightComp);
+
+		directionalLightComp->Direction = { -1.0f, 0.0f, 0.0f, 0.0f };
+		directionalLightComp->Color = { 212.0f / 255.0f, 175.0f / 255.0f, 55.0f / 255.0f, 0.25f };
```

Comment:
Suggestion/ preference: Might be good to use a neutral light (R==G==B) for preview of asset?

Reply:
I don't think the light makes much of a difference currently, haven't been able to resolve that